h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -539,6 +540,48 @@ protected int nFoldWork() {
     return N;
   }
 
+  protected transient LinkedBlockingQueue<ModelTrainingListener.Event> _events;
+  protected transient ModelTrainingListener _iterationListener;
+  protected transient ModelTrainingCoordinator _coordinator;
+
+  public class ModelTrainingCoordinator {
+
+    private final ModelBuilder<M, P, O>[] _cvModelBuilders;
+    private int _inProgress;
+
+
+    public ModelTrainingCoordinator(ModelBuilder<M, P, O>[] cvModelBuilders) {
+      _cvModelBuilders = cvModelBuilders;
+      _inProgress = _cvModelBuilders.length;
+    }
+
+    public void initStoppingParameters() {
+      cv_initStoppingParameters();
+    }
+    
+    public ModelTrainingCoordinator updateParameters() {
+      try {
+        while (_inProgress > 0) {
+          ModelTrainingListener.Event e = _events.take();
+          switch (e) {
+            case ALL_DONE:
+              _inProgress--;
+              break;
+            case ONE_DONE:
+              if (cv_updateOptimalParameters(_cvModelBuilders))
+                return this;
+              break;
+          }
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("TODO - better message", e);
+      }
+      cv_updateOptimalParameters(_cvModelBuilders);
+      return null;
```

Comment:
not sure it's necessary for `updateParameters` to return, even wondering if it's not a bug.
The only case where it returns `null` here is if the average `ntrees` decreases or remain constant.
I can't imagine this average decreasing (or the average epochs for DL for example), but it can remain "temporarily constant" because of the rounding (we're not even rounding, we're taking the floor value with `(int)`).
(3+3+3+3)/4 -> 3
(3+3+4+3)/4 -> 3, so we nullify the coordinator.
...
(3+3+5+5)/4 -> 4, coordinator was null, so it's not updated.

Reply:
It only returns `null` when all the CV models already finished = there is nothing to "coordinate" anymore.

If at least one CV is still in progress it waits for an `event` to arrive (it stays blocked).

If parameters don't change (in the case the number of trees won't grow in this case), method `cv_updateOptimalParameters` is supposed to return false and we wait for the next event to arrive. This continues until we have updated parameters all CV models finished and we have the final parameters.

@sebhrusen I don't see an issue there but maybe I am not understanding your comment properly

Reply:
correct! I got confused as `scoreAndBuildTrees` is called in 2 versions: one for CVs (with a listener), one for main model (with the coordinator), so yeah… it works fine.

One thing though: I'm still not sure that `updateParameters` should return anything.
It would work the same it method were `void`, and you could just reset `_coordinator` to null in the finally block of  `computeImpl`: I think it would avoid head scratching  when seeing this `_coordinator = _coordinator.updateParameters()`.

Or maybe I'm missing a point.

Reply:
It doesn't need to return anything in order to function properly. If that is confusing I can just remove it - I thought this way I would avoid accidentally callig `updateParameters` after they really cannot change anymore - but there is no harm right now, it will just not do anything.

If it is confusing I will remove it.